### ISSUE TITLE
Grad call (with no jit) is much slower than forward model call (that has some parts of it jitted) on cpu

### ISSUE THREAD

**Issue by kyrollosyanny** [google/jax#13658]
### Description

Hello, 

I have a forward model that consists of many function calls and if statements for control flow running on CPU. The computationally heavy parts of the forward model are jitted but the whole model is not because of the control flow. This makes the model fast to run **0.36 sec** per forward model call instead of 1.8 sec without jitting some functions in it.  

Now trying to call ```grad()```  on the forward model, takes much longer than the forward call itself.  It takes **47** seconds on the first grad call, and **6.2** seconds on subsequent grad calls. Note that I cannot use jit on grad since the forward model itself has some if statements.

Where would the slowdown be coming from? It doesnt seem to depend on the size of the input and the first call is much slower than subsequent calls. Finally, even the fastest grad call is slower than the version of the forward model with no jitted function. 

Thanks a lot for the help and pointers.  

### What jax/jaxlib version are you using?

pip install "jax[cuda11_cudnn82]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html

### Which accelerator(s) are you using?

CPU

### Additional system info

Python V 3.10.6. WSL.

**Comment by jakevdp** [comment]
It is expected that an un-jitted gradient of a oomputation will be slower than the computation itself.

Why? It's because the gradient in general is an intrinsically more complicated operation. For example, if you have the function $f(x) = x^n$, which is a single operation, the gradient will be $f^\prime(x) = 2 x^{n - 1}$, which is three operations.

You can see this play out by examining the jaxpr for this function and its gradient:
```python
import jax

def f(x):
  return x ** 2

print(jax.make_jaxpr(f)(10.0))
# { lambda ; a:f32[]. let b:f32[] = integer_pow[y=2] a in (b,) }

print(jax.make_jaxpr(jax.grad(f))(10.0))
# { lambda ; a:f32[]. let
#     _:f32[] = integer_pow[y=2] a
#     b:f32[] = integer_pow[y=1] a
#     c:f32[] = mul 2.0 b
#     d:f32[] = mul 1.0 c
#   in (d,) }
```
When you jit-compile the latter, XLA can often fuse these multiple operations to limit the slowdown. But the `grad` of a function almost always requires more operations than the function itself, so absent JIT it will almost always be a factor of a few slower.

So where does that leave you? If the cost of an un-jitted gradient is too much for your use case, I think the only way to do better is to JIT compile it, which will require modifying your code to make the original function compatible with JIT.

**Comment by kyrollosyanny** [comment]
That makes sense in terms of why it takes longer but I have a few quick follow-up questions:

1. In terms of restructuring the code, is the recommendation to use lax.cond for control-flow or to avoid conditional statements all together? If so, is there a best practice published somewhere?
2. I tried the same model in Pytorch, and while the gradient computation is slower than the forward model there as well. It is much closer in terms of time. For example, in Torch, the forward model takes 0.8 sec and the grad call takes 1 sec. As opposed to 0.3 sec forward call in Jax and 6 sec for grad?
3. Finally, is there a reason why the first grad call takes much longer than subsequent ones?

Thanks a lot for the help and explanations
Best ,

**Comment by jakevdp** [comment]
> 1. In terms of restructuring the code, is the recommendation to use lax.cond for control-flow or to avoid conditional statements all together? If so, is there a best practice published somewhere?

Yes, if the problem is `if` statements, you can generally replace those with `lax.cond`, `jnp.where`, masking, or another mechanism depending on the nature of the code.

> 2. I tried the same model in Pytorch, and while the gradient computation is slower than the forward model there as well. It is much closer in terms of time. For example, in Torch, the forward model takes 0.8 sec and the grad call takes 1 sec. As opposed to 0.3 sec forward call in Jax and 6 sec for grad?

It's hard to guess why the timings might be different without any hint of what your code is doing

> 3. Finally, is there a reason why the first grad call takes much longer than subsequent ones?

I suspect this is due to JIT compilation overhead (for the individual functions being called by grad) on the first run.
